File: examples/model-a/manual-multi-step/dotnet/ManualMultiStepInitiator.cs
using System.Text.Json.Nodes;
using Axme.Sdk;
using AxmeExamples;

namespace AxmeExamples.ModelA;

public class ManualMultiStepInitiator
{
    private static async Task<string> WaitForDone(AxmeClient client, string intentId, int timeoutSec = 60)
    {
        var since = 0; var deadline = DateTime.UtcNow.AddSeconds(timeoutSec);
        while (DateTime.UtcNow < deadline)
        {
            var events = await client.ListIntentEventsAsync(intentId, since);
            foreach (var evt in events["events"]?.AsArray() ?? [])
            {
                var st = evt?["status"]?.ToString() ?? "";
                if (evt?["seq"]?.GetValue<int>() is int seq) since = Math.Max(since, seq);
                if (st is "COMPLETED" or "IN_PROGRESS" or "FAILED" or "CANCELED" or "TIMED_OUT") return st;
            }
            await Task.Delay(1000);
        }
        return "UNKNOWN";
    }

    public static async Task Main(string[] args)
    {
        var baseUrl = SseHelper.Env("AXME_BASE_URL", "https://api.cloud.axme.ai");
        var apiKey = SseHelper.Env("AXME_API_KEY");
        var agent1 = SseHelper.Env("AXME_AGENT_1"); var agent2 = SseHelper.Env("AXME_AGENT_2");
        if (string.IsNullOrEmpty(apiKey)) { Console.Error.WriteLine("AXME_API_KEY is required"); return; }
        if (string.IsNullOrEmpty(agent1) || string.IsNullOrEmpty(agent2)) { Console.Error.WriteLine("AXME_AGENT_1 and AXME_AGENT_2 required"); return; }

        var client = new AxmeClient(new AxmeClientConfig { ApiKey = apiKey, BaseUrl = baseUrl });

        var payload = new JsonObject { ["change_id"] = "CHG-MULTI-STEP-001", ["service"] = "api-gateway",
            ["version"] = "5.0.0", ["environment"] = "staging", ["change_type"] = "config_update",
            ["risk_level"] = "medium", ["risk_threshold"] = 30 };

        // Step 1
        Console.WriteLine($"STEP 1: compliance check → {agent1}");
        var c1 = await client.CreateIntentAsync(new JsonObject { ["intent_type"] = "intent.compliance.check.v1",
            ["from_agent"] = "initiator://manual-multi-step", ["to_agent"] = agent1,
            ["correlation_id"] = Guid.NewGuid().ToString(), ["payload"] = payload.DeepClone() });
        var s1id = c1["intent_id"]?.ToString() ?? "";
        Console.WriteLine($"step 1 intent: {s1id} — waiting...");
        var s1 = await WaitForDone(client, s1id);
        Console.WriteLine($"step 1 → {s1}");

        if (s1 is not ("COMPLETED" or "IN_PROGRESS")) { Console.Error.WriteLine($"step 1 failed ({s1}) — aborting"); return; }

        // Step 2
        Console.WriteLine($"STEP 2: risk assessment → {agent2}");
        var p2 = payload.DeepClone().AsObject();
        p2["compliance_result"] = "passed"; p2["compliance_intent_id"] = s1id;
        var c2 = await client.CreateIntentAsync(new JsonObject { ["intent_type"] = "intent.risk.assessment.v1",
            ["from_agent"] = "initiator://manual-multi-step", ["to_agent"] = agent2,
            ["correlation_id"] = Guid.NewGuid().ToString(), ["payload"] = p2 });
        var s2id = c2["intent_id"]?.ToString() ?? "";
        Console.WriteLine($"step 2 intent: {s2id} — waiting...");
        var s2 = await WaitForDone(client, s2id);
        Console.WriteLine($"step 2 → {s2}");

        Console.WriteLine("\n=== Manual Multi-Step Pipeline ===");
        Console.WriteLine($"  Step 1 (compliance): {s1}  id={s1id}");
        Console.WriteLine($"  Step 2 (risk):       {s2}  id={s2id}");
        var ok = s1 is "COMPLETED" or "IN_PROGRESS" && s2 is "COMPLETED" or "IN_PROGRESS";
        Console.WriteLine($"  Result: {(ok ? "ALL STEPS PASSED" : "PIPELINE FAILED")}");
    }
}

File: examples/model-a/manual-multi-step/java/ManualMultiStepInitiator.java
package ai.axme.examples.modela;

import ai.axme.examples.SseHelper;
import dev.axme.sdk.*;
import java.util.*;

/** Model A — Manual Multi-Step (Java). */
public class ManualMultiStepInitiator {
    public static void main(String[] args) throws Exception {
        String apiKey = System.getenv("AXME_API_KEY");
        String baseUrl = SseHelper.env("AXME_BASE_URL", "https://api.cloud.axme.ai");
        String agent1 = System.getenv("AXME_AGENT_1"), agent2 = System.getenv("AXME_AGENT_2");
        if (apiKey == null || apiKey.isEmpty()) { System.err.println("AXME_API_KEY is required"); System.exit(1); }
        if (agent1 == null || agent2 == null || agent1.isEmpty() || agent2.isEmpty()) {
            System.err.println("AXME_AGENT_1 and AXME_AGENT_2 required"); System.exit(1);
        }

        AxmeClient client = new AxmeClient(new AxmeClientConfig(baseUrl, apiKey));
        Map<String, Object> payload = Map.of("change_id", "CHG-MULTI-STEP-001", "service", "api-gateway",
                "version", "5.0.0", "environment", "staging", "change_type", "config_update",
                "risk_level", "medium", "risk_threshold", 30);

        // Step 1
        System.out.printf("STEP 1: compliance check → %s%n", agent1);
        Map<String, Object> c1 = client.createIntent(Map.of("intent_type", "intent.compliance.check.v1",
                "from_agent", "initiator://manual-multi-step", "to_agent", agent1,
                "correlation_id", UUID.randomUUID().toString(), "payload", payload), RequestOptions.none());
        String s1id = (String) c1.get("intent_id");
        String s1 = waitForDone(client, s1id, 60_000);
        System.out.printf("step 1 → %s%n", s1);
        if (!"COMPLETED".equals(s1) && !"IN_PROGRESS".equals(s1)) { System.err.println("step 1 failed — aborting"); return; }

        // Step 2
        System.out.printf("STEP 2: risk assessment → %s%n", agent2);
        Map<String, Object> p2 = new HashMap<>(payload);
        p2.put("compliance_result", "passed"); p2.put("compliance_intent_id", s1id);
        Map<String, Object> c2 = client.createIntent(Map.of("intent_type", "intent.risk.assessment.v1",
                "from_agent", "initiator://manual-multi-step", "to_agent", agent2,
                "correlation_id", UUID.randomUUID().toString(), "payload", p2), RequestOptions.none());
        String s2id = (String) c2.get("intent_id");
        String s2 = waitForDone(client, s2id, 60_000);
        System.out.printf("step 2 → %s%n", s2);

        System.out.println("\n=== Manual Multi-Step Pipeline ===");
        System.out.printf("  Step 1 (compliance): %s  id=%s%n", s1, s1id);
        System.out.printf("  Step 2 (risk):       %s  id=%s%n", s2, s2id);
        boolean ok = ("COMPLETED".equals(s1) || "IN_PROGRESS".equals(s1)) && ("COMPLETED".equals(s2) || "IN_PROGRESS".equals(s2));
        System.out.println("  Result: " + (ok ? "ALL STEPS PASSED" : "PIPELINE FAILED"));
    }

    @SuppressWarnings("unchecked")
    static String waitForDone(AxmeClient client, String intentId, long timeoutMs) throws Exception {
        int since = 0; long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline) {
            Map<String, Object> events = client.listIntentEvents(intentId, since, RequestOptions.none());
            for (Map<String, Object> evt : (List<Map<String, Object>>) events.getOrDefault("events", List.of())) {
                String status = SseHelper.str(evt, "status");
                if (evt.get("seq") instanceof Number n) since = Math.max(since, n.intValue());
                if (Set.of("COMPLETED", "IN_PROGRESS", "FAILED", "CANCELED", "TIMED_OUT").contains(status)) return status;
            }
            Thread.sleep(1000);
        }
        return "UNKNOWN";
    }
}
